File: leetcode/59. Spiral Matrix II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int ii = 0, ie = n;
        int ji = 0, je = n;
        int i, j;
        vector<vector<int>> res(n, vector<int>(n, -1));
        int value = 1;
        while(value <= n*n){
            
            for(j = ji; j < je ; j++){
                res[ii][j] = value;
                value++;
            }
            ii++;
            for(i = ii; i < ie ; i++){
                res[i][je - 1] = value;
                value++;
            }
            je--;
            for(j = je - 1; j >= ji ; j--){
                res[ie - 1][j] = value;
                value++;
            }
            ie--;
            for(i = ie - 1; i >= ii ; i--){
                res[i][ji] = value;
                value++;
            }
            ji++;
        }
        
        return res;
    }
};

File: online-training-platform/leetcode/59. Spiral Matrix II.kt
class Solution {
    fun generateMatrix(n: Int): Array<IntArray> {

        if(n == 1){
            return Array<IntArray>(1){
                IntArray(1){
                    1
                }
            }
        }
        
        val cycle = (n / 2) + (n % 2)
        
        var rightFirst = 0; var rightSecond = 0
        var bottomFirst = 0; var bottomSecond = n - 1
        var leftFirst = n - 1; var leftSecond = n - 1
        var topFirst = n - 1; var topSecond = 0
        var value = 1
        var matrix = Array<IntArray>(n){
            IntArray(n){
                0
            }
        }
        repeat(cycle){
            val right = Pair(rightFirst, rightSecond)
            val bottom = Pair(bottomFirst, bottomSecond)
            val left = Pair(leftFirst, leftSecond)
            val top = Pair(topFirst, topSecond)
            

            for(j in rightSecond until bottomSecond) {
                if(value > n * n) break
                matrix[rightFirst][j] = value
                value += 1
            }

            for(i in bottomFirst until leftFirst){
                if(value > n * n) break
                matrix[i][bottomSecond] = value
                value += 1
            }

            var idx = leftSecond
            while( idx > topSecond) {
                if(value > n * n) break
                matrix[leftFirst][idx] = value
                idx -= 1
                value += 1
            }

            
            idx = topFirst
            while( idx > rightFirst) {
                if(value > n * n) break
                matrix[idx][topSecond] = value
                idx -= 1
                value += 1
            }

            rightFirst += 1
            rightSecond += 1

            bottomFirst += 1
            bottomSecond -= 1

            leftFirst -= 1
            leftSecond -= 1

            topFirst -= 1
            topSecond += 1


        }

        if(n % 2 == 1) {
            matrix[n/2][n/2] = n*n
        }
        

        return matrix
    }
}